sdk/core/azure-core/inc/azure/core/internal/log.hpp

```diff
@@ -3,9 +3,49 @@
 
 #pragma once
 
-#include "azure/core/logging/logging.hpp"
+#include "azure/core/dll_import_export.hpp"
+#include "azure/core/logger.hpp"
 
-namespace Azure { namespace Core { namespace Logging { namespace Internal {
-  bool ShouldLog(LogLevel level);
-  void Log(LogLevel level, std::string const& message);
-}}}} // namespace Azure::Core::Logging::Internal
+#include <atomic>
+#include <type_traits>
+
+namespace Azure { namespace Core { namespace Internal {
+  class Log {
+    using LogLevelInt = std::underlying_type<Logger::Level>::type;
+
+    static_assert(
+        std::is_same<int, LogLevelInt>::value == true && ATOMIC_INT_LOCK_FREE == 2,
+        "Logger::Level values must be representable as lock-free");
+
+    static_assert(ATOMIC_BOOL_LOCK_FREE == 2, "atomic<bool> must be lock-free");
+
+    static AZ_CORE_DLLEXPORT std::atomic<bool> g_isLoggingEnabled;
+    static AZ_CORE_DLLEXPORT std::atomic<LogLevelInt> g_logLevel;
+
+    static Logger::Listener GetLogListener();
+
+    static Logger::Listener GetLogListener(Logger::Level level)
+    {
+      return g_isLoggingEnabled && static_cast<LogLevelInt>(level) <= g_logLevel
+          ? GetLogListener()
+          : Logger::Listener(nullptr);
+    }
+
+    Log() = delete;
+    ~Log() = delete;
+
+  public:
+    static bool ShouldWrite(Logger::Level level) { return GetLogListener(level) != nullptr; }
+
+    static void Write(Logger::Level level, std::string const& message)
+    {
+      if (auto listener = GetLogListener(level))
+      {
+        listener(level, message);
+      }
+    }
+
+    static void ToggleLogging(bool isEnabled);
```

Comment:
Why do we need this?

Reply:
This is used internally by `Logger::SetListener()`:
```
void Logger::SetListener(Logger::Listener listener)
{
  std::unique_lock<std::shared_timed_mutex> loggerLock(g_logListenerMutex);
  g_logListener = std::move(listener);
  Log::ToggleLogging(g_logListener != nullptr);
}
```

ToggleLogging is inline function that sets an atomic boolean flag:
```
static AZ_CORE_DLLEXPORT std::atomic<bool> g_isLoggingEnabled;
static AZ_CORE_DLLEXPORT std::atomic<LogLevelInt> g_logLevel;
```

That boolean flag ends up being used by `Log::ShouldWrite()` inline:
```
static bool ShouldWrite(Logger::Level level) { return GetLogListener(level) != nullptr; }
```

`GetLogListener(Logger::Level level)` is inline and does lock-free check, based on that atomic boolean flag, and atomic logLevel.

```
static Logger::Listener GetLogListener(Logger::Level level)
    {
      return g_isLoggingEnabled && static_cast<LogLevelInt>(level) <= g_logLevel
          ? GetLogListener()
          : Logger::Listener(nullptr);
    }
```

Finally, `GetLogListener()` (without arguments) uses mutex and is non-inline.
```
Logger::Listener Log::GetLogListener()
{
  std::unique_lock<std::shared_timed_mutex> loggerLock(g_logListenerMutex);
  return g_logListener;
}
```

Reply:
I still don't understand. If you're already using the read/writer lock, then why is toggling necessary? Just get the shared lock when getting a copy of the value and use the exclusive lock when setting the value - this is all you need.